lib/workflow/stores/mongo.js

```diff
@@ -639,22 +639,39 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
         assert.uuid(data.taskId, 'data.taskId');
         assert.string(data.state, 'data.state');
 
-        var query = {
+        var queryDependency = {
             graphId: data.graphId,
-            reachable: true
+            reachable: true,
         };
-        query['dependencies.' + data.taskId] = {
+        queryDependency['dependencies.' + data.taskId] = {
             $in: [data.state, Constants.Task.States.Finished]
         };
-        var update = {
+        
+        var updateDependency = {
+            $unset: {}
+        };
+        updateDependency.$unset['dependencies.' + data.taskId] = '';
+
+        var queryAnyOf = {
+            graphId: data.graphId,
+            reachable: true,
+        };
+        queryAnyOf['dependencies.anyOf.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
+        var updateAnyOf = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
+        updateAnyOf.$unset["dependencies.anyOf"] = "";
+
         var options = {
             multi: true
         };
-
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        
+        return Promise.all([
```

Comment:
The promises within the `all` are executed concurrently, that means you could never expect the first promise is always executed before the 2nd one. The worse situation is that the `updateMongo` is not an atomic operation, so the 2nd updateMongo may be interrupted by  the 1st updateMongo.

For safely concern, these two promises should be executed serially, the [Promise.each](http://bluebirdjs.com/docs/api/promise.each.html) can ensure the serial order.